File: backend/app/routers/account.py
from __future__ import annotations

import io
import json
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from sqlalchemy import desc

from app.database import get_db
from app.models import Company, Report, User, UserWatchlistItem, UserFavoriteItem, Notification
from app.schemas import (
    CompanyOut, ReportOut, UserOut, NotificationOut, UpdateProfileRequest, UpdatePreferencesRequest,
)
from app.routers.auth import get_current_user
from app.services.notifications import load_notification_preferences, save_notification_preferences


router = APIRouter(prefix="/account", tags=["account"])


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _user_to_out(user: User) -> UserOut:
    return UserOut(
        id=user.id,
        email=user.email,
        full_name=user.full_name,
        investing_style=user.investing_style,
        theme_mode=user.theme_mode,
        accent_color=user.accent_color,
        dashboard_layout=user.dashboard_layout,
        card_density=user.card_density,
        ui_preferences=json.loads(user.ui_preferences_json or "{}"),
        notification_preferences=load_notification_preferences(user),
        is_active=user.is_active,
        google_connected=bool(user.google_sub),
    )


def _notification_to_out(notification: Notification) -> NotificationOut:
    return NotificationOut(
        id=notification.id,
        user_id=notification.user_id,
        company_id=notification.company_id,
        trigger_type=notification.trigger_type,
        channel=notification.channel,
        title=notification.title,
        body=notification.body,
        deep_link=notification.deep_link,
        metadata=json.loads(notification.metadata_json or "{}"),
        status=notification.status,
        read_at=notification.read_at,
        delivered_at=notification.delivered_at,
        created_at=notification.created_at,
    )


@router.get("/profile", response_model=UserOut)
def get_profile(current_user: User = Depends(get_current_user)):
    return _user_to_out(current_user)


@router.put("/profile", response_model=UserOut)
def update_profile(payload: UpdateProfileRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    if payload.full_name is not None:
        current_user.full_name = payload.full_name.strip() or None
    if payload.investing_style is not None:
        current_user.investing_style = payload.investing_style
    if payload.email is not None:
        email = payload.email.strip().lower()
        if email != current_user.email and db.query(User).filter(User.email == email).first():
            raise HTTPException(409, "An account with that email already exists")
        current_user.email = email
    db.commit()
    db.refresh(current_user)
    return _user_to_out(current_user)


@router.put("/preferences", response_model=UserOut)
def update_preferences(payload: UpdatePreferencesRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    if payload.theme_mode is not None:
        current_user.theme_mode = payload.theme_mode
    if payload.accent_color is not None:
        current_user.accent_color = payload.accent_color
    if payload.dashboard_layout is not None:
        current_user.dashboard_layout = payload.dashboard_layout
    if payload.card_density is not None:
        current_user.card_density = payload.card_density
    if payload.ui_preferences is not None:
        current_user.ui_preferences_json = json.dumps(payload.ui_preferences)
    if payload.notification_preferences is not None:
        save_notification_preferences(current_user, payload.notification_preferences)
    db.commit()
    db.refresh(current_user)
    return _user_to_out(current_user)


@router.get("/notifications", response_model=list[NotificationOut])
def get_notifications(limit: int = 50, unread_only: bool = False, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    query = db.query(Notification).filter(Notification.user_id == current_user.id)
    if unread_only:
        query = query.filter(Notification.read_at.is_(None))
    return [_notification_to_out(notification) for notification in query.order_by(desc(Notification.created_at), desc(Notification.id)).limit(limit).all()]


@router.post("/notifications/{notification_id}/read", response_model=NotificationOut)
def mark_notification_read(notification_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    notification = db.query(Notification).filter(Notification.id == notification_id, Notification.user_id == current_user.id).first()
    if not notification:
        raise HTTPException(404, "Notification not found")
    if notification.read_at is None:
        notification.read_at = _utcnow()
        db.commit()
        db.refresh(notification)
    return _notification_to_out(notification)


@router.post("/notifications/read-all")
def mark_all_notifications_read(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    updated = (
        db.query(Notification)
        .filter(Notification.user_id == current_user.id, Notification.read_at.is_(None))
        .update({Notification.read_at: _utcnow()}, synchronize_session=False)
    )
    db.commit()
    return {"message": "Notifications marked as read", "updated": updated}


@router.get("/watchlist", response_model=list[CompanyOut])
def get_watchlist(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    companies = (
        db.query(Company)
        .join(UserWatchlistItem, UserWatchlistItem.company_id == Company.id)
        .filter(UserWatchlistItem.user_id == current_user.id)
        .order_by(desc(UserWatchlistItem.created_at))
        .all()
    )
    return [CompanyOut.model_validate(company) for company in companies]


@router.post("/watchlist/{company_id}")
def add_watchlist_item(company_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    company = db.query(Company).filter(Company.id == company_id).first()
    if not company:
        raise HTTPException(404, "Company not found")
    existing = db.query(UserWatchlistItem).filter(UserWatchlistItem.user_id == current_user.id, UserWatchlistItem.company_id == company_id).first()
    if existing:
        return {"message": "Already in watchlist"}
    db.add(UserWatchlistItem(user_id=current_user.id, company_id=company_id))
    db.commit()
    return {"message": "Added to watchlist"}


@router.delete("/watchlist/{company_id}")
def remove_watchlist_item(company_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    item = db.query(UserWatchlistItem).filter(UserWatchlistItem.user_id == current_user.id, UserWatchlistItem.company_id == company_id).first()
    if item:
        db.delete(item)
        db.commit()
    return {"message": "Removed from watchlist"}


@router.get("/favorites", response_model=list[CompanyOut])
def get_favorites(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    companies = (
        db.query(Company)
        .join(UserFavoriteItem, UserFavoriteItem.company_id == Company.id)
        .filter(UserFavoriteItem.user_id == current_user.id)
        .order_by(desc(UserFavoriteItem.created_at))
        .all()
    )
    return [CompanyOut.model_validate(company) for company in companies]


@router.post("/favorites/{company_id}")
def add_favorite_item(company_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    company = db.query(Company).filter(Company.id == company_id).first()
    if not company:
        raise HTTPException(404, "Company not found")
    existing = db.query(UserFavoriteItem).filter(UserFavoriteItem.user_id == current_user.id, UserFavoriteItem.company_id == company_id).first()
    if existing:
        return {"message": "Already saved"}
    db.add(UserFavoriteItem(user_id=current_user.id, company_id=company_id))
    db.commit()
    return {"message": "Saved"}


@router.delete("/favorites/{company_id}")
def remove_favorite_item(company_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    item = db.query(UserFavoriteItem).filter(UserFavoriteItem.user_id == current_user.id, UserFavoriteItem.company_id == company_id).first()
    if item:
        db.delete(item)
        db.commit()
    return {"message": "Removed"}


@router.get("/reports", response_model=list[ReportOut])
def get_reports(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return (
        db.query(Report)
        .filter(Report.user_id == current_user.id)
        .order_by(desc(Report.uploaded_at), desc(Report.id))
        .all()
    )


@router.patch("/reports/{report_id}", response_model=ReportOut)
def rename_report(report_id: int, payload: dict, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    report = db.query(Report).filter(Report.id == report_id, Report.user_id == current_user.id).first()
    if not report:
        raise HTTPException(404, "Report not found")
    new_name = str(payload.get("file_name") or payload.get("name") or "").strip()
    if not new_name:
        raise HTTPException(400, "Report name cannot be empty")
    report.file_name = new_name
    db.commit()
    db.refresh(report)
    return report


@router.delete("/reports/{report_id}")
def delete_report(report_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    report = db.query(Report).filter(Report.id == report_id, Report.user_id == current_user.id).first()
    if not report:
        raise HTTPException(404, "Report not found")
    db.delete(report)
    db.commit()
    return {"message": "Report deleted"}


@router.get("/export")
def export_account(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    payload = {
        "profile": _user_to_out(current_user).model_dump(),
        "watchlist": [item.company_id for item in db.query(UserWatchlistItem).filter(UserWatchlistItem.user_id == current_user.id).all()],
        "favorites": [item.company_id for item in db.query(UserFavoriteItem).filter(UserFavoriteItem.user_id == current_user.id).all()],
        "reports": [ReportOut.model_validate(report).model_dump() for report in db.query(Report).filter(Report.user_id == current_user.id).all()],
    }
    buffer = io.BytesIO(json.dumps(payload, indent=2).encode("utf-8"))
    return StreamingResponse(buffer, media_type="application/json", headers={"Content-Disposition": f'attachment; filename="tricard-account-export.json"'})


@router.post("/import")
async def import_account(file: UploadFile = File(...), overwrite: bool = Query(False), db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    try:
        payload = json.loads((await file.read()).decode("utf-8"))
    except Exception:
        raise HTTPException(400, "Invalid JSON import file")

    if not isinstance(payload, dict):
        raise HTTPException(400, "Unsupported import format")

    if not overwrite and (payload.get("reports") or payload.get("watchlist") or payload.get("favorites")):
        raise HTTPException(409, "Import would overwrite existing data. Set overwrite=true to confirm.")

    profile = payload.get("profile", {})
    if isinstance(profile, dict):
        if profile.get("full_name") is not None:
            current_user.full_name = profile.get("full_name")
        if profile.get("investing_style") is not None:
            current_user.investing_style = profile.get("investing_style")
        if profile.get("theme_mode") is not None:
            current_user.theme_mode = profile.get("theme_mode")
        if profile.get("accent_color") is not None:
            current_user.accent_color = profile.get("accent_color")
        if profile.get("dashboard_layout") is not None:
            current_user.dashboard_layout = profile.get("dashboard_layout")
        if profile.get("card_density") is not None:
            current_user.card_density = profile.get("card_density")
        if isinstance(profile.get("ui_preferences"), dict):
            current_user.ui_preferences_json = json.dumps(profile.get("ui_preferences"))

    db.query(UserWatchlistItem).filter(UserWatchlistItem.user_id == current_user.id).delete()
    db.query(UserFavoriteItem).filter(UserFavoriteItem.user_id == current_user.id).delete()

    for company_id in payload.get("watchlist", []) or []:
        if db.query(Company).filter(Company.id == company_id).first():
            db.add(UserWatchlistItem(user_id=current_user.id, company_id=company_id))

    for company_id in payload.get("favorites", []) or []:
        if db.query(Company).filter(Company.id == company_id).first():
            db.add(UserFavoriteItem(user_id=current_user.id, company_id=company_id))

    db.commit()
    return {"message": "Account data imported"}

File: backend-node/src/routes/account.js
'use strict';

const express = require('express');
const prisma = require('../database');
const { requireAuth } = require('../middleware/auth');
const activityLogger = require('../utils/activityLogger');
const config = require('../config');
const ws = require('../websocket');
const notificationDispatcher = require('../services/notificationDispatcher');

const router = express.Router();

function safeJson(str) {
  try { return JSON.parse(str || '{}'); } catch { return {}; }
}

function userToOut(user) {
  const notifPrefs = safeJson(user.notificationPreferencesJson) ?? {};
  return {
    id: user.id,
    email: user.email,
    fullName: user.fullName,
    avatarUrl: user.avatarUrl,
    bio: user.bio,
    timezone: user.timezone,
    language: user.language,
    investingStyle: user.investingStyle,
    themeMode: user.themeMode,
    accentColor: user.accentColor,
    dashboardLayout: user.dashboardLayout,
    cardDensity: user.cardDensity,
    uiPreferences: safeJson(user.uiPreferencesJson),
    notificationPreferences: {
      enabled: true,
      emailAlerts: true,
      livePriceAlerts: true,
      priceMoveThresholdPct: 2.5,
      marketOpenCountries: ['Singapore', 'United States', 'Hong Kong'],
      ...notifPrefs,
    },
    isActive: user.isActive,
    isAdmin: user.isAdmin,
    emailVerified: user.emailVerified,
    googleConnected: Boolean(user.googleSub),
    twoFactorEnabled: user.twoFactorEnabled,
    createdAt: user.createdAt,
  };
}

// ── GET /account/profile ──────────────────────────────────

router.get('/profile', requireAuth, (req, res) => {
  res.json(userToOut(req.user));
});

router.get('/notification-channels', requireAuth, (req, res) => {
  res.json({
    telegram: {
      configured: Boolean(config.telegram.botToken && config.telegram.botUsername),
      botUsername: config.telegram.botUsername || null,
    },
    discord: { configured: true },
    email: { configured: Boolean(config.smtp.host) },
  });
});

// ── PUT /account/profile ──────────────────────────────────

router.put('/profile', requireAuth, async (req, res, next) => {
  try {
    const {
      fullName,
      full_name: fullNameSnake,
      investingStyle,
      investing_style: investingStyleSnake,
      email,
      bio,
      timezone,
      language,
    } = req.body;
    const updates = {};

    if (fullName !== undefined || fullNameSnake !== undefined) {
      updates.fullName = (fullName ?? fullNameSnake)?.trim() || null;
    }
    if (investingStyle !== undefined || investingStyleSnake !== undefined) {
      updates.investingStyle = investingStyle ?? investingStyleSnake;
    }
    if (bio !== undefined) updates.bio = bio?.trim() || null;
    if (timezone !== undefined) updates.timezone = timezone;
    if (language !== undefined) updates.language = language;

    if (email !== undefined) {
      const normalized = email.trim().toLowerCase();
      if (normalized !== req.user.email) {
        const conflict = await prisma.user.findUnique({ where: { email: normalized } });
        if (conflict) return res.status(409).json({ detail: 'Email already in use.' });
        updates.email = normalized;
        updates.emailVerified = false;
      }
    }

    const updated = await prisma.user.update({ where: { id: req.user.id }, data: updates });
    activityLogger.log(req.user.id, 'update_profile', 'user', req.user.id, {}, req);
    res.json(userToOut(updated));
  } catch (err) {
    next(err);
  }
});

// ── PUT /account/preferences ──────────────────────────────

router.put('/preferences', requireAuth, async (req, res, next) => {
  try {
    const {
      themeMode,
      theme_mode: themeModeSnake,
      accentColor,
      accent_color: accentColorSnake,
      dashboardLayout,
      dashboard_layout: dashboardLayoutSnake,
      cardDensity,
      card_density: cardDensitySnake,
      uiPreferences,
      ui_preferences: uiPreferencesSnake,
      notificationPreferences,
      notification_preferences: notificationPreferencesSnake,
    } = req.body;
    const updates = {};

    if (themeMode ?? themeModeSnake) updates.themeMode = themeMode ?? themeModeSnake;
    if (accentColor ?? accentColorSnake) updates.accentColor = accentColor ?? accentColorSnake;
    if (dashboardLayout ?? dashboardLayoutSnake) updates.dashboardLayout = dashboardLayout ?? dashboardLayoutSnake;
    if (cardDensity ?? cardDensitySnake) updates.cardDensity = cardDensity ?? cardDensitySnake;
    if (uiPreferences ?? uiPreferencesSnake) updates.uiPreferencesJson = JSON.stringify(uiPreferences ?? uiPreferencesSnake);
    if (notificationPreferences ?? notificationPreferencesSnake) updates.notificationPreferencesJson = JSON.stringify(notificationPreferences ?? notificationPreferencesSnake);

    const updated = await prisma.user.update({ where: { id: req.user.id }, data: updates });
    res.json(userToOut(updated));
  } catch (err) {
    next(err);
  }
});

// ── Avatar upload ──────────────────────────────────────────

const multer = require('multer');
const path = require('path');
const fs = require('fs');

const avatarStorage = multer.diskStorage({
  destination: (req, file, cb) => {
    const dir = path.join(config.uploadDir, 'avatars');
    fs.mkdirSync(dir, { recursive: true });
    cb(null, dir);
  },
  filename: (req, file, cb) => {
    const ext = path.extname(file.originalname) || '.jpg';
    cb(null, `user_${req.user.id}${ext}`);
  },
});

const avatarUpload = multer({
  storage: avatarStorage,
  limits: { fileSize: 5 * 1024 * 1024 },
  fileFilter: (req, file, cb) => {
    if (!file.mimetype.startsWith('image/')) return cb(new Error('Only image files are allowed.'));
    cb(null, true);
  },
});

router.post('/avatar', requireAuth, avatarUpload.single('avatar'), async (req, res, next) => {
  try {
    if (!req.file) return res.status(400).json({ detail: 'No file uploaded.' });
    const avatarUrl = `/uploads/avatars/${req.file.filename}`;
    await prisma.user.update({ where: { id: req.user.id }, data: { avatarUrl } });
    res.json({ avatarUrl });
  } catch (err) {
    next(err);
  }
});

// ── Notifications ──────────────────────────────────────────

router.get('/notifications', requireAuth, async (req, res, next) => {
  try {
    const { limit = '50', unreadOnly = 'false' } = req.query;
    const where = { userId: req.user.id };
    if (unreadOnly === 'true') where.readAt = null;

    const notifications = await prisma.notification.findMany({
      where,
      orderBy: [{ createdAt: 'desc' }, { id: 'desc' }],
      take: parseInt(limit, 10),
    });
    res.json(notifications.map(n => ({ ...n, metadata: safeJson(n.metadataJson) })));
  } catch (err) {
    next(err);
  }
});

router.post('/notifications/:id/read', requireAuth, async (req, res, next) => {
  try {
    const notification = await prisma.notification.findFirst({
      where: { id: parseInt(req.params.id), userId: req.user.id },
    });
    if (!notification) return res.status(404).json({ detail: 'Notification not found.' });

    if (!notification.readAt) {
      await prisma.notification.update({ where: { id: notification.id }, data: { readAt: new Date() } });
    }
    const updated = await prisma.notification.findUnique({ where: { id: notification.id } });
    res.json({ ...updated, metadata: safeJson(updated.metadataJson) });
  } catch (err) {
    next(err);
  }
});

router.post('/notifications/read-all', requireAuth, async (req, res, next) => {
  try {
    const { count } = await prisma.notification.updateMany({
      where: { userId: req.user.id, readAt: null },
      data: { readAt: new Date() },
    });
    res.json({ message: 'All notifications marked as read.', updated: count });
  } catch (err) {
    next(err);
  }
});

router.delete('/notifications/:id', requireAuth, async (req, res, next) => {
  try {
    const notification = await prisma.notification.findFirst({
      where: { id: parseInt(req.params.id), userId: req.user.id },
    });
    if (!notification) return res.status(404).json({ detail: 'Notification not found.' });
    await prisma.notification.delete({ where: { id: notification.id } });
    res.json({ message: 'Notification deleted.' });
  } catch (err) {
    next(err);
  }
});

// ── Watchlist ──────────────────────────────────────────────

router.get('/watchlist', requireAuth, async (req, res, next) => {
  try {
    const items = await prisma.userWatchlistItem.findMany({
      where: { userId: req.user.id },
      include: { company: true },
      orderBy: { createdAt: 'desc' },
    });
    res.json(items.map(i => i.company));
  } catch (err) {
    next(err);
  }
});

router.post('/watchlist/:companyId', requireAuth, async (req, res, next) => {
  try {
    const companyId = parseInt(req.params.companyId);
    const company = await prisma.company.findUnique({ where: { id: companyId } });
    if (!company) return res.status(404).json({ detail: 'Company not found.' });

    const existing = await prisma.userWatchlistItem.findUnique({
      where: { userId_companyId: { userId: req.user.id, companyId } },
    });
    if (existing) return res.json({ message: 'Already in watchlist.', created: false });

    await prisma.userWatchlistItem.create({ data: { userId: req.user.id, companyId } });
    const notification = await prisma.notification.create({
      data: {
        userId: req.user.id,
        companyId,
        triggerType: 'watchlist_added',
        channel: 'IN_APP',
        title: 'Added to watchlist',
        body: `${company.name}${company.ticker ? ` (${company.ticker})` : ''} is now being tracked.`,
        deepLink: `/#/companies/${companyId}`,
        deliveredAt: new Date(),
      },
    });
    const delivery = await notificationDispatcher.dispatchInAppNotification({
      user: req.user,
      title: notification.title,
      body: notification.body,
      triggerType: notification.triggerType,
      companyName: company.name,
    });
    await prisma.notification.update({
      where: { id: notification.id },
      data: { metadataJson: JSON.stringify({ delivery }) },
    });
    ws.pushToUser(req.user.id, { notification });
    res.json({ message: 'Added to watchlist.', created: true, notification, delivery });
  } catch (err) {
    next(err);
  }
});

router.delete('/watchlist/:companyId', requireAuth, async (req, res, next) => {
  try {
    await prisma.userWatchlistItem.deleteMany({
      where: { userId: req.user.id, companyId: parseInt(req.params.companyId) },
    });
    res.json({ message: 'Removed from watchlist.' });
  } catch (err) {
    next(err);
  }
});

// ── Favorites ──────────────────────────────────────────────

router.get('/favorites', requireAuth, async (req, res, next) => {
  try {
    const items = await prisma.userFavoriteItem.findMany({
      where: { userId: req.user.id },
      include: { company: true },
      orderBy: { createdAt: 'desc' },
    });
    res.json(items.map(i => i.company));
  } catch (err) {
    next(err);
  }
});

router.post('/favorites/:companyId', requireAuth, async (req, res, next) => {
  try {
    const companyId = parseInt(req.params.companyId);
    const company = await prisma.company.findUnique({ where: { id: companyId } });
    if (!company) return res.status(404).json({ detail: 'Company not found.' });

    await prisma.userFavoriteItem.upsert({
      where: { userId_companyId: { userId: req.user.id, companyId } },
      update: {},
      create: { userId: req.user.id, companyId },
    });
    res.json({ message: 'Added to favorites.' });
  } catch (err) {
    next(err);
  }
});

router.delete('/favorites/:companyId', requireAuth, async (req, res, next) => {
  try {
    await prisma.userFavoriteItem.deleteMany({
      where: { userId: req.user.id, companyId: parseInt(req.params.companyId) },
    });
    res.json({ message: 'Removed from favorites.' });
  } catch (err) {
    next(err);
  }
});

// ── Company Tags (Extra Feature 5) ────────────────────────

router.get('/tags', requireAuth, async (req, res, next) => {
  try {
    const tags = await prisma.userCompanyTag.findMany({
      where: { userId: req.user.id },
      include: { company: true },
      orderBy: { createdAt: 'desc' },
    });
    res.json(tags);
  } catch (err) {
    next(err);
  }
});

router.post('/tags/:companyId', requireAuth, async (req, res, next) => {
  try {
    const { tag } = req.body;
    if (!tag) return res.status(400).json({ detail: 'Tag is required.' });
    const companyId = parseInt(req.params.companyId);

    await prisma.userCompanyTag.upsert({
      where: { userId_companyId_tag: { userId: req.user.id, companyId, tag } },
      update: {},
      create: { userId: req.user.id, companyId, tag },
    });
    res.json({ message: 'Tag added.', tag });
  } catch (err) {
    next(err);
  }
});

router.delete('/tags/:companyId', requireAuth, async (req, res, next) => {
  try {
    const { tag } = req.body;
    const companyId = parseInt(req.params.companyId);
    const where = { userId: req.user.id, companyId };
    if (tag) where.tag = tag;

    await prisma.userCompanyTag.deleteMany({ where });
    res.json({ message: 'Tag(s) removed.' });
  } catch (err) {
    next(err);
  }
});

// ── Activity log ──────────────────────────────────────────

router.get('/activity', requireAuth, async (req, res, next) => {
  try {
    const { limit = '50' } = req.query;
    const logs = await prisma.activityLog.findMany({
      where: { userId: req.user.id },
      orderBy: { createdAt: 'desc' },
      take: parseInt(limit, 10),
    });
    res.json(logs);
  } catch (err) {
    next(err);
  }
});

// ── Export account data (GDPR-style) ──────────────────────

router.get('/export', requireAuth, async (req, res, next) => {
  try {
    const [user, watchlist, favorites, notifications, activityLogs] = await Promise.all([
      prisma.user.findUnique({ where: { id: req.user.id } }),
      prisma.userWatchlistItem.findMany({ where: { userId: req.user.id }, include: { company: true } }),
      prisma.userFavoriteItem.findMany({ where: { userId: req.user.id }, include: { company: true } }),
      prisma.notification.findMany({ where: { userId: req.user.id } }),
      prisma.activityLog.findMany({ where: { userId: req.user.id } }),
    ]);

    const exportData = {
      exportedAt: new Date().toISOString(),
      profile: userToOut(user),
      watchlist: watchlist.map(i => i.company),
      favorites: favorites.map(i => i.company),
      notifications: notifications.length,
      activityLogs: activityLogs.length,
    };

    res.setHeader('Content-Type', 'application/json');
    res.setHeader('Content-Disposition', 'attachment; filename="my-esg-data.json"');
    res.json(exportData);
  } catch (err) {
    next(err);
  }
});

// ── Reports (uploaded documents) ──────────────────────────

router.get('/reports', requireAuth, async (req, res, next) => {
  try {
    const reports = await prisma.report.findMany({
      where: { userId: req.user.id },
      orderBy: { uploadedAt: 'desc' },
    });
    res.json(reports);
  } catch (err) {
    next(err);
  }
});

router.patch('/reports/:id', requireAuth, async (req, res, next) => {
  try {
    const id = parseInt(req.params.id, 10);
    const { file_name } = req.body;
    const report = await prisma.report.findFirst({ where: { id, userId: req.user.id } });
    if (!report) return res.status(404).json({ detail: 'Report not found' });
    const updated = await prisma.report.update({ where: { id }, data: { fileName: file_name } });
    res.json(updated);
  } catch (err) {
    next(err);
  }
});

router.delete('/reports/:id', requireAuth, async (req, res, next) => {
  try {
    const id = parseInt(req.params.id, 10);
    const report = await prisma.report.findFirst({ where: { id, userId: req.user.id } });
    if (!report) return res.status(404).json({ detail: 'Report not found' });
    await prisma.report.delete({ where: { id } });
    res.json({ success: true });
  } catch (err) {
    next(err);
  }
});

module.exports = router;
